### PR TITLE
Jetpack Manage: Adding pagination to the sites v2 dashboard

### PR DESCRIPTION
**Pull request by coder-karen**
Related to https://github.com/Automattic/jetpack-manage/issues/311

## Proposed Changes

* This PR adds pagination to the sites V2 dashboard.
* This is assumes a starting per page figure of 50, which was pre-inputted, but can be changed if needed.
* This PR also includes field label translations, although not related to the linked issue

## Testing Instructions

* To test pagination, connect multiple Jurassic Ninja sites to Jetpack (that's assuming you don't already have 10 or more sites in the Sites dashboard).
* You will also need to apply the WPcom patch at D140045-code, make sure you are sandboxed and your hosts file points to WPcom (until or unless that patch gets merged)
* Then with this PR applied, visit the Sites dashboard: `http://jetpack.cloud.localhost:3000/sites`
* From the 'View options' icon (top right) of the sites dashboard, select 10 rows per page. You should see 10 connected sites showing per page, and on selecting page 2 from the pagination options (bottom right), the remaining connected sites should show. If you have more than 20 sites, you should have more pagination options.
* If you then select 20 rows per page or a number higher than your number of sites, then all sites should display in the dashboard.

**Screenshots**

Pagination selector (if you don't have the development environment label and bug report badge in the way...)
Screenshot 2024-02-28 at 13 19 42

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by cleacos** [comment]
Also, for testing, you can set the initial `perPage` state to `2`:

```
{
		type: 'table',
		perPage: 2,
		page: currentPage,
		sort: {
			field: 'url',
			direction: 'desc',
		},
		search: search,
		filters:
			filter?.issueTypes?.map( ( issueType ) => {
				return {
					field: 'status',
					operator: 'in',
					value: filtersMap.find( ( filterMap ) => filterMap.filterType === issueType )?.ref || 1,
				};
			} ) || [],
		hiddenFields: [ 'status' ],
		layout: {},
		selectedSite: undefined,
	}
```

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx [hunk 1]
```diff
@@ -23,7 +23,13 @@ const SitesDataViews = ( {
...
 	const translate = useTranslate();
 
-	const sites = useFormattedSites( data?.sites ?? [] );
+	const totalSites = data ? data.sites.length : 0;
+	const sitesPerPage = sitesViewState.perPage;
+	const totalPages = Math.ceil( totalSites / sitesPerPage );
+	const pageEndPoint = sitesViewState.page * sitesPerPage;
+	const pageStartPoint = pageEndPoint - ( sitesPerPage - 1 );
+	const unformattedSites = data ? data.sites.slice( pageStartPoint - 1, pageEndPoint ) : [];
+	const sites = useFormattedSites( unformattedSites );
```
Comment (tbradsha) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: A few points:

* `totalSites` assumes `data.sites` exists. Previously we didn't have this assumption. Does it always exist?
* `sitesPerPage`: I'd suggest requiring sitesPerPage to be `>0` in case something is stored wrong on the backend. `0` would give a div-by-zero error, and negatives would throw all the math off.
* `unformattedSites` also has the assumption that `data.sites` exists.

As for `pageStartPoint`, you're adding 1 to its value, and then you subtract one on the only time you use it in the `slice`. Perhaps just set its value to `pageEndPoint - sitesPerPage`?

Given that arrays are zero-indexed, in their current states `pageStartPoint` is really "first site plus one", and `pageEndPoint` is really "next first site".

**Review by tbradsha — CHANGES_REQUESTED** [review]
On the surface this seems to work well:

* 10 or fewer: doesn't show
* 10: shows and navigates between the sites.

There are a few areas where I could see errors cropping up depending on how reliable the data is. I've left those comments inline.

The biggest issue I see, though, is that we're filtering the site data we retrieve from the endpoint, so would be limited to the number of sites that endpoint returns. Instead, we should be retrieving the sites from the endpoint on each page change.

Two other things that may or may not be in scope here:

* We can't change the per-page preference
* The URL doesn't change when switching pages.

**Comment by cleacos** [comment]
**I found an issue:** If I go to the page, let's say, page 3, and then I click on `Favorites` and the total favorite sites fit in page 1, it will result in a `No results` page (without pagination).

If the Sites submenu is selected, we should reset any active pagination.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx (hunk 1 above)
Comment (gogdzl) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: Mainly agree with Thomas.
Didn't test, but some suggestions:
```suggestion
	const sitesPerPage = Math.max( 1, sitesViewState.perPage );
	const totalPages = Math.ceil( totalSites / sitesPerPage );
	const pageEndPoint = sitesViewState.page * sitesPerPage;
	const pageStartPoint = Math.max( 1, pageEndPoint - sitesPerPage + 1 );
	const unformattedSites = data?.sites?.slice( pageStartPoint - 1, pageEndPoint ) || [];
	const sites = useFormattedSites( unformattedSites );
```

Comment (tbradsha) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: Great! Though I'd tweak it to this (untested):

```
	const pageStartPoint = pageEndPoint - sitesPerPage;
	const unformattedSites = data?.sites?.slice( pageStartPoint, pageEndPoint ) || [];
```

Comment (coder-karen) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: In the end, only needed `totalPages`, `totalSites` and `sitesPerPage`.

**Comment by coder-karen** [comment]
> The biggest issue I see, though, is that we're filtering the site data we retrieve from the endpoint, so would be limited to the number of sites that endpoint returns. Instead, we should be retrieving the sites from the endpoint on each page change.

@tbradsha  Thank you for flagging this. Once the relevant diff is deployed this should be fixed (though it can be tested before hand as well - see testing instructions above).

> We can't change the per-page preference
> The URL doesn't change when switching pages.

Both, I believe, are out of scope (since pagination doesn't affect the URL, and the dataviews default behaviour is that the per-page selection is cleared on moving away from that page - we'd have to discuss further about whether to save that information somewhere other than locally if we want to do that).


> I found an issue: If I go to the page, let's say, page 3, and then I click on Favorites and the total favorite sites fit in page 1, it will result in a No results page (without pagination).

@cleacos Great catch. This should reset now (page number) when visiting those pages.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx [hunk 2]
```diff
@@ -23,13 +23,10 @@ const SitesDataViews = ( {
 }: SitesDataViewsProps ) => {
 	const translate = useTranslate();
 
-	const totalSites = data ? data.sites.length : 0;
-	const sitesPerPage = sitesViewState.perPage;
+	const totalSites = data ? data?.total : 0;
+	const sitesPerPage = sitesViewState.perPage > 0 ? sitesViewState.perPage : 10;
```
Comment (tbradsha) [inline client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: Tiny nitpick: the backend default is 20, so perhaps we should match that?

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts [hunk 3]
```diff
@@ -45,6 +53,7 @@ const useFetchDashboardSites = (
 					...( currentPage && { page: currentPage } ),
 					...agencyDashboardFilterToQueryObject( filter ),
 					...agencyDashboardSortToQueryObject( sort ),
+					per_page: per_page ?? null,
```
Comment (tbradsha) [inline client/data/agency-dashboard/use-fetch-dashboard-sites.ts]: Perhaps we should pass the per_page default if `null`/`undefined`?

**Comment by tbradsha** [comment]
Looking over the code, it seems to pretty good. A tiny nitpick was mentioned above (different defaults on frontend vs. backend).

Two other things to consider:
* We always pass the `per_page` `GET` param, but by default it's blank. I'd recommend passing the actual default value.
* The new Sites dash uses the `per_page` set in the JS side rather than the one returned by the API. This can result in odd states (e.g. no pagination) when the `per_page` is manually manipulated. This is a way-to-the-edge case, though.

The only blocker at the moment is the unit tests.

File: client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx (hunk 2 above)
Comment (coder-karen) [reply client/jetpack-cloud/sections/agency-dashboard/sites-overview/sites-dataviews/index.tsx]: Yes, oversight at my end, thanks!

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts (hunk 3 above)
Comment (coder-karen) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts]: I had that initially, my only reservation was that if the backend default changes, we won't necessarily know to change it here. But then, this all may be somewhat more irrelevant when looking at a new API and implementation for A4A. And it's the same issue for `totalPages` in `SitesDataViews` anyway :D So I'll change it.

**Comment by tbradsha** [comment]
As discussed in a call:

* We need to fix the unit tests
* Pagination for favorites isn't quite there

**Review by tbradsha — APPROVED** [review]
Tests pass and pagination works. 🚀

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/11514197

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @coder-karen for including a screenshot in the description! This is really helpful for our translators.

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts [hunk 4]
```diff
@@ -30,10 +30,24 @@ const useFetchDashboardSites = (
...
+		per_page,
+	];
+	// If per_page is not provided, we want to remove per_page from the query_key as existing tests don't pass otherwise.
+	if ( ! per_page ) {
+		query_key = [ 'jetpack-agency-dashboard-sites', searchQuery, currentPage, filter, sort ];
+	}
+
 	return useQuery( {
-		queryKey: [ 'jetpack-agency-dashboard-sites', searchQuery, currentPage, filter, sort ],
+		queryKey: query_key,
```
Comment (cleacos) [inline client/data/agency-dashboard/use-fetch-dashboard-sites.ts:50]: A small improvement to not use `let` and that condition would have been:

```
const queryKey = [
		'jetpack-agency-dashboard-sites',
		searchQuery,
		currentPage,
		filter,
		sort,
		...( per_page ? [ per_page ] : [] ),
	];

	return useQuery( {
		queryKey,
```

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts [hunk 5]
```diff
@@ -45,6 +59,7 @@ const useFetchDashboardSites = (
 					...( currentPage && { page: currentPage } ),
 					...agencyDashboardFilterToQueryObject( filter ),
 					...agencyDashboardSortToQueryObject( sort ),
+					per_page: per_page ?? 20,
```
Comment (cleacos) [inline client/data/agency-dashboard/use-fetch-dashboard-sites.ts:62]: `...( per_page && { per_page: per_page ?? 20 } ),` ?

Comment (cleacos) [inline client/data/agency-dashboard/use-fetch-dashboard-sites.ts:62]: `...( per_page && { per_page } ),` ?

**Comment by cleacos** [comment]
I created this PR to collect and update minor improvements: https://github.com/Automattic/wp-calypso/pull/88073

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts (hunk 4 above)
Comment (coder-karen) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts:50]: We experimented with that a little earlier, the current way was (so far) the only way to prevent an eslint warning: 		'The following dependencies are missing in your queryKey: per_page'

File: client/data/agency-dashboard/use-fetch-dashboard-sites.ts (hunk 5 above)
Comment (coder-karen) [reply client/data/agency-dashboard/use-fetch-dashboard-sites.ts:62]: This also works, so can be included in an update, thanks.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11514197).
